Extend functionality of degree function

emlyn opened the pull request:
Extend the `degree` function to:
- allow degrees larger than one octave, such as a [thirteenth](https://en.wikipedia.org/wiki/Thirteenth)
- allow [augmented and diminished](https://en.wikipedia.org/wiki/Interval_(music)#Augmented_and_diminished) degrees by prefixing with an `a`/`d`, and doubly augmented/diminished with `aa`/ `dd`, as well `p` for perfect intervals, for completeness.
- allow numeric strings/symbols as well as roman numerals, so you can use `'d5'`/`:d5` as well as `:dv` for diminished 5th

The documentation will also need updating, but I wanted to check if this looks reasonable first.

emlyn commented:
I realise that `from_roman` supporting numbers up to >4000 is a slight overkill, but I wasn't sure where to draw the line (maybe only supporting digits `i`, `v` & `x` would be reasonable?). Also it might make sense to move it somewhere more general in case it's useful elsewhere.
There is a slight ambiguity in that `d` is both a prefix for "diminished", and the number 500, but that's unlikely to be a problem in practise.

samaaron commented:
Looks fab, thanks!

Is there any chance you could update the documentation appropriately? Thanks also for the tests :-)

emlyn commented:
It wasn't obvious to me how to get all the relevant info concisely into the documentation... how does this look?

emlyn commented:
Please continue to be a pain :-) I'm happy if it makes this better!

I'll think of a couple of examples and add them later.

samaaron commented:
Ah, sorry for not being clearer. This is via a weird doc system I created just for Sonic Pi so there's no wonder it's non-obvious ;-)

There's essentially a big metadata map for each public function and one of the keys of this map is       `examples:` which has a list as a value with each element in the list being a separate example. The funky part is that the comments will be automatically moved to the right hand side when formatting the documentation for viewing, so it's possible to copy the code independently of the descriptive comments. 

Here's an example: https://github.com/samaaron/sonic-pi/blob/master/app/server/ruby/lib/sonicpi/lang/sound.rb#L3615

emlyn commented:
Thanks for pointing that out, I had noticed the examples before, but forgot. I've removed the "E.g." part from the doc string and added a couple of examples.

I hope they aren't too complicated, I wanted to include an example of using it to create a chord that's not built in to Sonic Pi.

ethancrawford commented:
Just want to check that we're ok with showing the bracket syntax for blocks `{...}` here @samaaron given that it's not officially documented as part of the Sonic Pi DSL? Looks great anyway @emlyn!

samaaron commented:
Agree with @ethancrawford, the following doesn't have any direct instruction in the tutorial:

```
[:i, :iii, :v, :dvii, :dix, :Axi, :xiii].map {|d| (degree d, :Fs, :major)}
```

Whilst it's valid ruby, it's unsupported Sonic Pi code as nowhere do we cover mapping functions over lists.

emlyn commented:
Yes, that's a fair point. Is there a supported way of doing something similar? What about:
```ruby
[:i, :iii, :v, :dvii, :dix, :Axi, :xiii].each do |d|
  play (degree d, :Fs, :major)
end
```
although it's not quite the same, as it won't scale the volume according to the number of notes.
I could always remove the loop entirely:
```ruby
play [(degree :i, :Fs, :major), (degree :iii, :Fs, :major), (degree :v, :Fs, :major),
      (degree :dvii, :Fs, :major), (degree :dix, :Fs, :major), (degree :Axi, :Fs, :major),
      (degree :xiii, :Fs, :major)]
```
but I think that's a bit too verbose/repetitive.

emlyn commented:
How about this, would it be acceptable? Or else do you have any suggestions on how to improve it?

ethancrawford commented:
🤔 hmmm
Well, *technically* speaking, `.each` is only ever used a few times in 'documentation', and I think only in the example compositions - not in the tutorial or references per se. `append` on the other hand is pure ruby that is definitely not used anywhere user facing at all as far as I can tell. (Though I do see that you have added comments to try to mitigate these facts).
As for a better solution, I'm not 100% sure. Any ideas @samaaron ?

samaaron commented:
Thank-you :-)

ethancrawford commented:
@samaaron I take it you were ok with the pure ruby functions in this PR, as described in my comment above?

samaaron commented:
@ethancrawford I might still tweak them before release, but the PR is too good to not pull in :-)
